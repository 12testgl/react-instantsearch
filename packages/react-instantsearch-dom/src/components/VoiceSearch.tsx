import React from 'react';
import { translatable, Translate } from 'react-instantsearch-core';
import { createClassNames } from '../core/utils';
const cx = createClassNames('VoiceSearch');

type VoiceListeningState = {
  status: string;
  transcript?: string;
  isSpeechFinal?: boolean;
  errorCode?: string;
};

type ToggleListening = (toggleListeningParams: {
  searchAsYouSpeak: boolean;
}) => void;

type InnerComponentProps = {
  status: string;
  errorCode?: string;
  isListening: boolean;
  transcript?: string;
  isSpeechFinal?: boolean;
  isBrowserSupported: boolean;
};

type VoiceSearchProps = {
  isBrowserSupported: boolean;
  isListening: boolean;
  toggleListening: ToggleListening;
  voiceListeningState: VoiceListeningState;
  searchAsYouSpeak?: boolean;

  translate: Translate;
  buttonComponent?: React.FunctionComponent<InnerComponentProps>;
  statusComponent?: React.FunctionComponent<InnerComponentProps>;
};

const DefaultButton: React.FunctionComponent<InnerComponentProps> = ({
  status,
  errorCode,
  isListening,
}) => {
  if (status === 'error' && errorCode === 'not-allowed') {
    return (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        width="16"
        height="16"
        viewBox="0 0 24 24"
        fill="none"
        stroke="currentColor"
        strokeWidth="2"
        strokeLinecap="round"
        strokeLinejoin="round"
      >
        <line x1="1" y1="1" x2="23" y2="23" />
        <path d="M9 9v3a3 3 0 0 0 5.12 2.12M15 9.34V4a3 3 0 0 0-5.94-.6" />
        <path d="M17 16.95A7 7 0 0 1 5 12v-2m14 0v2a7 7 0 0 1-.11 1.23" />
        <line x1="12" y1="19" x2="12" y2="23" />
        <line x1="8" y1="23" x2="16" y2="23" />
      </svg>
    );
  } else if (isListening) {
    return (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        width="16"
        height="16"
        viewBox="0 0 24 24"
        fill="none"
        stroke="currentColor"
        strokeWidth="2"
        strokeLinecap="round"
        strokeLinejoin="round"
      >
        <path
          d="M12 1a3 3 0 0 0-3 3v8a3 3 0 0 0 6 0V4a3 3 0 0 0-3-3z"
          fill="currentColor"
        />
        <path d="M19 10v2a7 7 0 0 1-14 0v-2" />
        <line x1="12" y1="19" x2="12" y2="23" />
        <line x1="8" y1="23" x2="16" y2="23" />
      </svg>
    );
  } else {
    return (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        width="16"
        height="16"
        viewBox="0 0 24 24"
        fill="none"
        stroke="currentColor"
        strokeWidth="2"
        strokeLinecap="round"
        strokeLinejoin="round"
      >
        <path d="M12 1a3 3 0 0 0-3 3v8a3 3 0 0 0 6 0V4a3 3 0 0 0-3-3z" />
        <path d="M19 10v2a7 7 0 0 1-14 0v-2" />
        <line x1="12" y1="19" x2="12" y2="23" />
        <line x1="8" y1="23" x2="16" y2="23" />
      </svg>
    );
  }
};

const DefaultStatusComponent: React.FC<InnerComponentProps> = ({
  status,
}) => <p>{status}</p>;

const VoiceSearch = ({
  translate,
  isBrowserSupported,
  isListening,
  toggleListening,
  voiceListeningState,
  searchAsYouSpeak = false,

  buttonComponent: Button = DefaultButtonComponent,
  statusComponent: StatusComponent = DefaultStatusComponent,
}: VoiceSearchProps) => {
  const onClick = (event: React.MouseEvent<HTMLElement>) => {
    event.currentTarget.blur();
    toggleListening({ searchAsYouSpeak });
  };

  const { status, transcript, isSpeechFinal, errorCode } = voiceListeningState;

  const innerProps = {
    status,
    errorCode,
    isListening,
    transcript,
    isSpeechFinal,
    isBrowserSupported,
  };

  return (
    <div className={cx('')}>
      <button
        className={cx('button')}
        type="button"
        title={translate('buttonTitle')}
        onClick={onClick}
        disabled={!isBrowserSupported}
      >
        <Button {...innerProps} />
      </button>
      <div className={cx('div')}>
        <StatusComponent {...innerProps} />
      </div>
    </div>
  );
};

export default translatable({
  buttonTitle: 'Search by voice',
})(VoiceSearch);
