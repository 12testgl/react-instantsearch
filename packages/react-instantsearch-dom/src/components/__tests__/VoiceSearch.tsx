import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import VoiceSearch from '../VoiceSearch';

Enzyme.configure({ adapter: new Adapter() });

const renderAndMatchSnapshot = (element: JSX.Element) => {
  const instance = renderer.create(element);
  expect(instance.toJSON()).toMatchSnapshot();
  instance.unmount();
};

const defaultProps = {
  isBrowserSupported: true,
  isListening: false,
  toggleListening: () => {},
  voiceListeningState: {
    status: 'initial',
    transcript: undefined,
    isSpeechFinal: undefined,
    errorCode: undefined,
  },
};

describe('VoiceSearch', () => {
  it('does nothing', () => {
    expect(true).toBe(true);
  });

  describe('button', () => {
    it('calls toggleListening when button is clicked', () => {
      const props = {
        ...defaultProps,
        toggleListening: jest.fn(),
      };
      const wrapper = mount(<VoiceSearch {...props} />);
      wrapper.find('button').simulate('click');
      expect(props.toggleListening).toHaveBeenCalledTimes(1);
    });
  });

  describe('Rendering', () => {
    it('with default props', () => {
      renderAndMatchSnapshot(<VoiceSearch {...defaultProps} />);
    });

    it('with custom component for button with isListening: false', () => {
      const customButton = ({ isListening }) => (
        <button>{isListening ? 'Stop' : 'Start'}</button>
      );

      renderAndMatchSnapshot(
        <VoiceSearch {...defaultProps} buttonComponent={customButton} />
      );
    });

    it('with custom component for button with isListening: true', () => {
      const customButton = ({ isListening }) => (
        <button>{isListening ? 'Stop' : 'Start'}</button>
      );

      const props = {
        isBrowserSupported: true,
        isListening: true,
        toggleListening: () => {},
        voiceListeningState: {
          status: 'recognizing',
          transcript: undefined,
          isSpeechFinal: undefined,
          errorCode: undefined,
        },
      };

      renderAndMatchSnapshot(
        <VoiceSearch {...props} buttonComponent={customButton} />
      );
    });

    it('with custom template for status', () => {
      const customStatus = ({
        status,
        errorCode,
        isListening,
        transcript,
        isSpeechFinal,
        isBrowserSupported,
      }) => (
        <div>
          <p>status: {status}</p>
          <p>errorCode: {errorCode}</p>
          <p>isListening: {isListening}</p>
          <p>transcript: {transcript}</p>
          <p>isSpeechFinal: {isSpeechFinal}</p>
          <p>isBrowserSupported: {isBrowserSupported}</p>
        </div>
      );

      const props = {
        isBrowserSupported: true,
        isListening: true,
        toggleListening: () => {},
        voiceListeningState: {
          status: 'recognizing',
          transcript: 'Hello',
          isSpeechFinal: false,
          errorCode: undefined,
        },
      };

      renderAndMatchSnapshot(
        <VoiceSearch {...props} statusComponent={customStatus} />
      );
    });
  });
});
