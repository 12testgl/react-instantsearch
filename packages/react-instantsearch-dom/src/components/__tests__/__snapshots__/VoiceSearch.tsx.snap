// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoiceSearch Rendering with custom component for button with isListening: false 1`] = `
<div
  className="ais-VoiceSearch"
>
  <button
    className="ais-VoiceSearch-button"
    disabled={false}
    onClick={[Function]}
    title="Search by voice"
    type="button"
  >
    <button>
      Start
    </button>
  </button>
  <div
    className="ais-VoiceSearch-div"
  >
    <p>
      initial
    </p>
  </div>
</div>
`;

exports[`VoiceSearch Rendering with custom component for button with isListening: true 1`] = `
<div
  className="ais-VoiceSearch"
>
  <button
    className="ais-VoiceSearch-button"
    disabled={false}
    onClick={[Function]}
    title="Search by voice"
    type="button"
  >
    <button>
      Stop
    </button>
  </button>
  <div
    className="ais-VoiceSearch-div"
  >
    <p>
      recognizing
    </p>
  </div>
</div>
`;

exports[`VoiceSearch Rendering with custom template for status 1`] = `
<div
  className="ais-VoiceSearch"
>
  <button
    className="ais-VoiceSearch-button"
    disabled={false}
    onClick={[Function]}
    title="Search by voice"
    type="button"
  >
    <svg
      fill="none"
      height="16"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 1a3 3 0 0 0-3 3v8a3 3 0 0 0 6 0V4a3 3 0 0 0-3-3z"
        fill="currentColor"
      />
      <path
        d="M19 10v2a7 7 0 0 1-14 0v-2"
      />
      <line
        x1="12"
        x2="12"
        y1="19"
        y2="23"
      />
      <line
        x1="8"
        x2="16"
        y1="23"
        y2="23"
      />
    </svg>
  </button>
  <div
    className="ais-VoiceSearch-div"
  >
    <div>
      <p>
        status: 
        recognizing
      </p>
      <p>
        errorCode: 
      </p>
      <p>
        isListening: 
      </p>
      <p>
        transcript: 
        Hello
      </p>
      <p>
        isSpeechFinal: 
      </p>
      <p>
        isBrowserSupported: 
      </p>
    </div>
  </div>
</div>
`;

exports[`VoiceSearch Rendering with default props 1`] = `
<div
  className="ais-VoiceSearch"
>
  <button
    className="ais-VoiceSearch-button"
    disabled={false}
    onClick={[Function]}
    title="Search by voice"
    type="button"
  >
    <svg
      fill="none"
      height="16"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 1a3 3 0 0 0-3 3v8a3 3 0 0 0 6 0V4a3 3 0 0 0-3-3z"
      />
      <path
        d="M19 10v2a7 7 0 0 1-14 0v-2"
      />
      <line
        x1="12"
        x2="12"
        y1="19"
        y2="23"
      />
      <line
        x1="8"
        x2="16"
        y1="23"
        y2="23"
      />
    </svg>
  </button>
  <div
    className="ais-VoiceSearch-div"
  >
    <p>
      initial
    </p>
  </div>
</div>
`;
